src/Tester/TestStreamProviders/Generator/GeneratorPooledCache.cs

```diff
@@ -0,0 +1,197 @@
+            {
+                if (bufferPool == null)
+                {
+                    throw new ArgumentNullException("bufferPool");
+                }
+                if (purgeAction == null)
+                {
+                    throw new ArgumentNullException("purgeAction");
+                }
+                this.bufferPool = bufferPool;
+                this.purgeAction = purgeAction;
+            }
+
+            public void QueueMessageToCachedMessage(ref CachedMessage cachedMessage, GeneratedBatchContainer queueMessage)
+            {
+                cachedMessage.StreamGuid = queueMessage.StreamGuid;
+                cachedMessage.StreamNamespace = queueMessage.StreamNamespace;
+                cachedMessage.SequenceNumber = queueMessage.RealToken.SequenceNumber;
+                cachedMessage.Payload = SerializeMessageIntoPooledSegment(queueMessage);
+            }
+
+            // Placed object message payload into a segment from a buffer pool.  When this get's too big, older blocks will be purged
+            private ArraySegment<byte> SerializeMessageIntoPooledSegment(GeneratedBatchContainer queueMessage)
+            {
+                // serialize payload
+                byte[] serializedPayload = SerializationManager.SerializeToByteArray(queueMessage.Payload);
+                int size = serializedPayload.Length;
+
+                // get segment from current block
+                ArraySegment<byte> segment;
+                if (currentBuffer == null || !currentBuffer.TryGetSegment(size, out segment))
+                {
+                    // no block or block full, get new block and try again
+                    currentBuffer = bufferPool.Allocate();
+                    currentBuffer.SetPurgeAction(purgeAction);
+                    // if this fails with clean block, then requested size is too big
+                    if (!currentBuffer.TryGetSegment(size, out segment))
+                    {
+                        string errmsg = String.Format(CultureInfo.InvariantCulture,
+                            "Message size is to big. MessageSize: {0}", size);
+                        throw new ArgumentOutOfRangeException("queueMessage", errmsg);
+                    }
+                }
+                Buffer.BlockCopy(serializedPayload, 0, segment.Array, segment.Offset, size);
+                return segment;
+            }
+
+            public IBatchContainer GetBatchContainer(ref CachedMessage cachedMessage)
+            {
+                //Deserialize payload
+                var stream = new BinaryTokenStreamReader(cachedMessage.Payload.ToArray());
+                object payloadObject = SerializationManager.Deserialize(stream);
+                return new GeneratedBatchContainer(cachedMessage.StreamGuid, cachedMessage.StreamNamespace,
+                    payloadObject, new EventSequenceToken(cachedMessage.SequenceNumber));
```

Comment:
As a side note, these two sequence tokens happen to be the same for this implementation, but they will not always be the same.  The sequence token in the batch is the token to be delivered to the application, while the sequence token returned by GetSequenceToken is used for ordering in the cache, which represents the order of data in the queue.  These may not always be the same.  Consider a queue that has duplicates of application layer data.

Application sends event 1,2,3 gets a timeout on 3, and resends.  The first send of 3 that timed out actually succeeded, so the queue has application layer data 1,2,3,3, but queue ordering of 1,2,3,4.  The sequence token provided to the cache must reflect queue ordering (1,2,3,4), but when sending to the application, which will perform deduplication base off the sequence token, tokens should be 1,2,3,3 so that the second 3 will be detected as a duplicate and ignored.